### ISSUE TITLE
Using `useGlobals` inside DocPage throws an error: Cannot read property 'updateGlobals' of undefined

### ISSUE THREAD

**Issue by petmat** [storybookjs/storybook#12982]
**Describe the bug**
I need to access the globals set by toolbar addon in the custom DocPage I have set for the docs addon.
The toolbar addon documentation instructs to use the `useGlobals` hook to consume the globals within an addon.
However, when I try to use the `useGlobals` hook in the DocPage it throws an error:
```
Uncaught TypeError: Cannot read property 'updateGlobals' of undefined
    at useGlobals (index.js:550)
    at page (button.stories.tsx:67)
    at renderWithHooks (react-dom.development.js:16260)
    at mountIndeterminateComponent (react-dom.development.js:18794)
    at beginWork$1 (react-dom.development.js:20162)
    at HTMLUnknownElement.callCallback (react-dom.development.js:336)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:385)
    at invokeGuardedCallback (react-dom.development.js:440)
    at beginWork$$1 (react-dom.development.js:25780)
    at performUnitOfWork (react-dom.development.js:24698)
```

**To Reproduce**

1. Add a custom doc page globally or on component or story level
2. Use the `useGlobals` hook in the custom docs page
3. Go to storybook and the docs tab
4. Check the browser dev tools console for the error

**Expected behavior**
Expected to render the doc page without errors and have access to the globals.

**Code snippets**
Here's the code from my `DocsPage.tsx` file:
```ts
import * as React from "react";

import {
  ArgsTable,
  Description,
  PRIMARY_STORY,
  Primary,
  Stories,
  Subtitle,
  Title,
} from "@storybook/addon-docs/blocks";

import { useGlobals } from "@storybook/api";

export const DocsPage = () => {
  const [globals] = useGlobals();

  return (
      <Title />
      <Subtitle />
      <Description />
      <Primary />
      <ArgsTable story={PRIMARY_STORY} />
      <Stories />
  );
};
```
And here's how I set the doc page in `preview.ts` file with `addParameters` function:
```ts
addParameters({
  viewport,
  options: { storySort, theme },
  controls: { hideNoControlsWarning: true },
  docs: {
    page: DocsPage
  },
});
```

**System**
Environment Info:

  System:
    OS: macOS Mojave 10.14.6
    CPU: (12) x64 Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
  Binaries:
    Node: 12.18.2 - ~/.nvm/versions/node/v12.18.2/bin/node
    Yarn: 1.22.4 - ~/.yarn/bin/yarn
    npm: 6.14.5 - ~/.nvm/versions/node/v12.18.2/bin/npm
  Browsers:
    Chrome: 86.0.4240.111
    Firefox: 72.0.2
    Safari: 14.0

**Additional Info**
I'm using `@storybook/core@6.0.27` and `@storybook/addon-essentials@6.0.27`.

**Comment by petmat** [comment]
After examining the source code I've gone to the conclusion that `useGlobals` should be used when making a addon to give it access to the global values. So it's probably not expected to work out-of-the-box in the custom doc page.

But the question remains that is there a way to access globals in the doc page?

**Comment by shilman** [comment]
AFAIK Storybook hooks are only available inside stories/decorators cc @Hypnosphi

**Comment by Hypnosphi** [comment]
@shilman that's correct

**Comment by petmat** [comment]
Thank you for your reply! The reason I wanted to access globals in doc page was to set background color styles based on the selected theme. Since it's not possible in doc page, I ended up setting some global CSS styles targeting the docs CSS classes in a story decorator. Not the prettiest solution but it works.

So I managed to work around it 👍   I'm not sure how big of a problem this is. Go ahead and close the issue if you feel it's not that big of a deal.

**Comment by shilman** [comment]
@tmeasday seems like we need a whole new set of hooks for docs. we hacked this for args. WDYT about globals?

**Comment by tmeasday** [comment]
Seems like if we do one we should do the other 🤷

**Comment by williamliangwl** [comment]
Wondering if this is going to be added ?

I want to create an addon on Toolbar that affect the story rendering of both Story and Docs, so I plan to have the "communication" done through Globals

Or is there a workaround for that use case ?

**Comment by frassinier** [comment]
@williamliangwl  Here is a workaround I found today!

```js
export const globalTypes = {
	theme: {
		name: 'Theme',
		description: 'Choose a theme to apply',
		toolbar: {
			icon: 'paintbrush',
			items: [
				{ value: 'light', left: '⚪️', title: 'Default theme' },
				{ value: 'dark', left: '⚫️', title: 'Dark theme' },
			],
		},
	},
};

export const parameters = {
	docs: {
		container: props => {
			const channel = addons.getChannel();
			const hasDarkMode = props.context.globals.theme === 'dark';
			return (
				<>
					<Switch
						label={'Dark mode'}
						onChange={() => {
							channel.emit(UPDATE_GLOBALS, {
								globals: { theme: hasDarkMode ? 'light' : 'dark' },
							});
						}}
						checked={hasDarkMode}
					/>
					<DocsContainer {...props} />
				</>
			);
		},
	},
};
```

Hope it can help!

**Comment by pauledwards241** [comment]
Thanks @frassinier, we've just hit the same issue and have been able to work around it with an adapted version of your code 👍

**Comment by AntonisFK** [comment]
Is there any updates on this?

**Comment by taletski** [comment]
+1 
My use case: need to controll theme from stories level to create visual testing stories for all themes.

**Comment by noranda** [comment]
I'd also like to see this. I have a theme switcher and want to show a color palette based on the selected theme.

**Comment by tmeasday** [comment]
So, there are a couple things here.

1. In the original reproduction, @petmat is importing `useGlobals` from `@storybook/api`, which is the manager API (the package has been renamed in 7.0), and docs runs in the preview. So it isn't surprising it doesn't work.

2. There is a preview side `useGlobals` also, you can now import from `@storybook/preview-api` or `@storybook/addons`. However, that's not useable outside of stories or decorators.

3. We actually have a doc blocks `useGlobals` that we use internally: https://github.com/storybookjs/storybook/blob/7d9bb5495e94324a9bea278383793c2f0f4a70e9/code/ui/blocks/src/blocks/useGlobals.ts#L6-L20 -- but AFAICT it's not currently exported from `@storybook/blocks`. @JReinhold do you think we should?

In the meantime, there's nothing stopping you from writing similar code to the above in your `DocPage`, which is basically what @frassinier does.

**Comment by JReinhold** [comment]
> 3. We actually have a doc blocks `useGlobals` that we use internally: [`7d9bb54`/code/ui/blocks/src/blocks/useGlobals.ts#L6-L20](https://github.com/storybookjs/storybook/blob/7d9bb5495e94324a9bea278383793c2f0f4a70e9/code/ui/blocks/src/blocks/useGlobals.ts#L6-L20) -- but AFAICT it's not currently exported from `@storybook/blocks`. @JReinhold do you think we should?

I can see how that can be valuable to have to users, yeah. However I think it should be able to change globals too, and not just read them. doesn't sound too complicated though.

**Comment by tmeasday** [comment]
Right. And there's no real reason why it needs to take a `PreparedStory` as an argument (which you might not have). So there's a bit of work to clean this up to be more useful. Maybe a `good first issue`? Should we use this ticket or make a new one?

**Comment by shilman** [comment]
Let's clean up this `useGlobals` function per @tmeasday and @JReinhold 's suggestions and export it. It will solve the problem for everybody here and also for #22563

**Comment by dolanmiu** [comment]
For our use case, we need useGlobals because we have images which are illegible with dark mode enabled.

We're trying to use the theme to switch between the images in a purpose built React component

AFAIK, storybook doesn't support GitHub's solution to use #gh-light-mode-only at the end of images?

**Comment by LeonxLJX** [comment]
I'd like to take this one.

I've read through the report. Before changing anything I'd like to reproduce it against current `main` first — my reasoning is that a fix anchored to the actual observed behavior is much easier to review than one inferred from the report alone, especially if the surrounding code has moved since this was filed.

Plan: reproduce and confirm the root cause, then make the smallest change that addresses it (plus a regression test if the repo has coverage for this area). If the root cause turns out to live somewhere other than where the report points, I'll report back here before forcing a change.

Anything already known about this that I should read first?

(claiming via @LeonxLJX)
